framework/src/modules/chain/dpos/account.js

```diff
@@ -0,0 +1,239 @@
+const { getAddressFromPublicKey } = require('@liskhq/lisk-cryptography');
+const BigNum = require('@liskhq/bignum');
+
+class Account {
+	constructor({
+		storage,
+		slots,
+		activeDelegates,
+		logger,
+		delegates,
+		exceptions,
+	}) {
+		this.storage = storage;
+		this.slots = slots;
+		this.activeDelegates = activeDelegates;
+		this.logger = logger;
+		this.delegates = delegates;
+		this.exceptions = exceptions;
+	}
+
+	async apply(block, tx) {
+		return this._update(block, false, tx);
+	}
+
+	async undo(block, tx) {
+		return this._update(block, true, tx);
+	}
+
+	/**
+	 * @param {Block} block
+	 */
+	async _update(block, undo, tx) {
+		const round = this.slots.calcRound(block.height);
+
+		await this._updateProducedBlocks(block, undo, tx);
+
+		if (this._isLastBlockOfTheRound(block)) {
+			const summarizedRound = await this._summarizeRound(block, tx);
+			const nonForgedDelegateAddresses = await this._getNonForgedDelegateAddresses(
+				round,
+				summarizedRound.delegatePks,
+			);
+			await this._updateMissedBlocks(nonForgedDelegateAddresses, tx);
+			const updatedDelegates = await this._distributeRewardsAndFees(
+				summarizedRound,
+				undo,
+				tx,
+			);
+			await this._updateVotes(updatedDelegates, undo, tx);
+		}
+	}
+
+	async _updateProducedBlocks(block, undo, tx) {
+		const address = getAddressFromPublicKey(block.generatorPublicKey);
+		const filters = { address_eq: address };
```

Comment:
You can skip the `getAddressFromPublicKey` step and search by `publicKey` directly